runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java

```diff
@@ -464,16 +472,32 @@ private void setNextWatermarkTimer(StreamingRuntimeContext runtime) {
 
   /** Visible so that we can check this in tests. Must not be used for anything else. */
   @VisibleForTesting
-  public List<? extends UnboundedSource<OutputT, CheckpointMarkT>> getLocalSplitSources() {
+  List<? extends UnboundedSource<OutputT, CheckpointMarkT>> getLocalSplitSources() {
     return localSplitSources;
   }
 
   /** Visible so that we can check this in tests. Must not be used for anything else. */
   @VisibleForTesting
-  public List<UnboundedSource.UnboundedReader<OutputT>> getLocalReaders() {
+  List<UnboundedSource.UnboundedReader<OutputT>> getLocalReaders() {
     return localReaders;
   }
 
+  /** Visible so that we can check this in tests. Must not be used for anything else. */
+  @VisibleForTesting
+  boolean isRunning() {
+    return isRunning;
+  }
+
+  /**
```

Comment:
I would prefer we use `Whitebox::setInternalState` instead of adding an extra method. Same could be done for `isRunning` above.

Reply:
I would use Whitebox only in case of private methods/fields where the scope cannot be changed. Here we can control the scope. The method is package-private, thus is not user-facing. Using Whitebox is very fragile in case of any refactoring, whereas the package-private method will be tied to the flag and can be fixed immediately.

Generally speaking, we don't want extra test methods, but since it is hidden for the user, already common practice throughout the code base, annotated as a test method, and read-only, I think it is perfectly acceptable.

Reply:
We are cluttering main code with things that are only necessary for tests. That's the reason why more sophisticated test libraries like Mockito are around. 

As for fragility: I would agree if main code relies on reflection. But this is a different situation. The test will break when things are changed in the class that it tests, and that is expected. And it will show as soon as you run it.

Reply:
I agree that code should be structured for testability. For example it should be possible to test classes in isolation. But since that isn't always practically possible, libraries that allow testing without obstructing the main code were built.

This is a topic of debate in general. I also used to be skeptical of using Mockito in general and in this way specifically. But after some time and learning how it really helps towards the goal of having reliable and concise code and that in any way you run your tests as part if development, this is no longer a concern to me. 

When I change the internals of a class, then I run the test for that class. That will tell me immediately that the field no longer exists and it is obvious what needs to be adjusted since it happens in the same context. This would be the reality of working with less type-safe languages also - you know if your change is good or not by running tests.

Reply:
I suppose it comes down to the preferred workflow. I like the simplicity of the package-private approach. I can be sure that I don't break tests for a simple refactoring. If you can run tests very easily and fast (which is a bit harder in our build setup), it doesn't really matter. Of course in the end you always run tests, but the less magic involved, the less surprises you will get.

>I also used to be skeptical of using Mockito in general and in this way specifically.

I took the opposite turn. I grew skeptical of Mockito although I tried to use it whenever possible before ;) Without a dobut, it's a very useful tool.